src/main/java/com/bootcamp/demo/repository/FirebaseAbstractRepository.java

```diff
@@ -0,0 +1,90 @@
+package com.bootcamp.demo.repository;
+
+import com.google.api.core.ApiFuture;
+import com.google.cloud.firestore.*;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.ParameterizedType;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+import com.google.cloud.firestore.CollectionReference;
+import org.apache.commons.beanutils.PropertyUtils;
+
+public abstract class FirebaseAbstractRepository<T> implements IRepository<T> {
+    private final Class<T> className;
+
+    @SuppressWarnings("unchecked")
+    public FirebaseAbstractRepository() {
+        this.className = ((Class<T>) (
+                (ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0]);
+    }
+
+    public abstract CollectionReference getCollection();
+
+
+    public Set<T> findAll() {
+        try {
+            Iterable<QueryDocumentSnapshot> documents = getCollection()
+                    .get()
+                    .get()
+                    .getDocuments();
+            Set<T> result = new HashSet<>();
+            for (QueryDocumentSnapshot doc : documents) {
+                T object = doc.toObject(className);
+                PropertyUtils.setProperty(object, "id", doc.getId());
+                result.add(object);
+            }
+            return result;
+        } catch (ExecutionException | InterruptedException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            throw new RuntimeException();
```

Comment:
Why do we throw a RuntimeException here? Maybe it would help to define a specific exception (e.g. a RepositoryException), so that the service or other classes that use the repository to be able to add specific error handling.

Reply:
I heard that RuntimeException are preffered. The idea of a specific exception was on the back of my mind from the very beginning, but I postponed and then I digressed from the original plan.
Hopes this resolve the issue, also I m open to change it further if required.
Fixed: 16aa2c55696e8a22d266bfc1ecc7c3711b0c2421